### PR TITLE
[TRTLLM-9676][fix] Fix mamba_cache_manager when enabling cuda_graph_padding and let test cover this case

### PR DESCRIPTION
**Pull request by JadoTu**
## Summary by CodeRabbit

* **Improvements**
  * Enhanced request padding optimization for Mamba models in CUDA Graph execution mode, improving performance consistency and stability across variable batch sizes.
  * Justification: [MambaCacheManage.pptx](https://nvidia.sharepoint.com/:p:/r/sites/Jian_Tu_Space/_layouts/15/doc2.aspx?sourcedoc=%7B01B41782-8F22-4D24-8DE7-61381AE4AA1C%7D&file=MambaCacheManager.pptx&action=edit&mobileredirect=true)

* **Tests**
  * Expanded CUDA Graph accuracy testing with dynamic batch size support spanning 1 to 2048, plus improved sampling strategies for validation.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

-->

## Description

1. Current MambaCacheManager will not deal with padding requests' indices of ssm states. Repeated indices will allure errors in CUDA kernels. This PR is to remove the repeatance.
2. Change the CI test in order to cover this scenario.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by nv-guomingz — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py [hunk 1]
```diff
@@ -450,6 +451,12 @@ def _get_padded_batch(self, batch: ScheduledRequests,
             if spec_res_mgr:
                 spec_res_mgr.add_dummy_requests([CUDA_GRAPH_DUMMY_REQUEST_ID])
 
+        # handle special cases of padding requests + MambaCacheManager or MambaHybridCacheManager
+        if isinstance(kv_cache_manager,
+                      (MambaCacheManager, MambaHybridCacheManager)):
```
Comment (yuantailing) [inline tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py]: Could remove `MambaHybridCacheManager` because it derives `MambaCacheManager`.

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py [hunk 2]
```diff
@@ -126,6 +134,18 @@ def _prepare_mamba_cache_blocks(self, request_ids: List[int]):
...
             state_indices, dtype=torch.int32, device=self.ssm_states.device)
 
+    # When there exists padded requests, the state indices should not be repeated.
+    def reorder_state_indices_when_padding_requests(self, request_size,
+                                                    padding_size):
+        if padding_size == 0:
+            return
+
+        self.request_mask[:] = True
+        self.request_mask[self.state_indices[:request_size]] = False
```
Comment (yuantailing) [inline tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py]: There are implicit CUDA synchronizations which negate the benefits of the overlap scheduler.

**Review by kmk142789 — APPROVED** [review]

File: tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py (hunk 1 above)
Comment (JadoTu) [reply tensorrt_llm/_torch/pyexecutor/cuda_graph_runner.py]: fixed in new patch. Thanks.

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py (hunk 2 above)
Comment (yuantailing) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py]: Here are synchronizations I said. The NVTX range "====A1==" is this code snippet I quoted.

image

There is no synchronization when I delete the code snippet. But my NVTX trace is generated before your new method.

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py [hunk 3]
```diff
@@ -127,6 +127,23 @@ def _prepare_mamba_cache_blocks(self, request_ids: List[int]):
...
+        assert len(
+            self.mamba_cache_free_blocks
+        ) >= padding_size, "Padding requests run out of available mamba cache blocks"
+        self.state_indices[request_size:request_size +
+                           padding_size] = torch.tensor(
+                               self.mamba_cache_free_blocks[:padding_size],
+                               dtype=self.state_indices.dtype,
+                               pin_memory=True).to(self.state_indices.device,
+                                                   non_blocking=True)
+
```
[the remark this answers is not in the document]
Comment (JadoTu) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py]: Thanks for the inputs. I will try to figure out.

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 4]
```diff
@@ -4660,6 +4662,7 @@ def test_bf16_4gpu(self, tp_size, pp_size, ep_size, cuda_graph,
             task.evaluate(llm)
             mocker.patch.object(GSM8K, "MAX_OUTPUT_LEN",
                                 self.GSM8K_MAX_OUTPUT_LEN)
+            mocker.patch.object(GSM8K, "NUM_SAMPLES", 1319)
```
[the remark this answers is not in the document]
Comment (JadoTu) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py]: I want to make sure here is 1319, even if someone else change the default sample numbers. The total samples run is important to assert the correctness of cuda graph padding.

[the remark this answers is not in the document]
Comment (nv-guomingz) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py]: Either we call GSM8K.NUM_SAMPLES directly or we add a assertion here.

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 5]
```diff
@@ -4642,10 +4642,12 @@ def test_bf16_4gpu(self, tp_size, pp_size, ep_size, cuda_graph,
...
                                         enable_block_reuse=False)
-        pytorch_config = dict(disable_overlap_scheduler=not overlap_scheduler,
-                              cuda_graph_config=CudaGraphConfig(
-                                  max_batch_size=512, enable_padding=True)
-                              if cuda_graph else None)
+        pytorch_config = dict(
+            disable_overlap_scheduler=not overlap_scheduler,
+            cuda_graph_config=CudaGraphConfig(
+                enable_padding=True,
+                batch_sizes=[1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048])
```
[the remark this answers is not in the document]
Comment (JadoTu) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py]: This bug relies on specific cuda graph batchsize setting. But it makes no sense here if the default setting is the same when setting `max_batch_size = 2048` 😆

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuantailing) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py]: The runtime error is resolved, but there are CUDA syncs.

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py (hunk 2 above)
Comment (yuantailing) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py]: The runtime error has been resolved, but there are CUDA syncs.

Comment (yuantailing) [reply tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py]: The sync issue has been resolved.

File: tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py [hunk 6]
```diff
@@ -109,23 +109,59 @@ def __init__(
...
+        ), "Padding requests run out of available mamba cache blocks"
+        # we can use mamba_cache_free_blocks for padding_requests
+        if padding_size <= len(self.mamba_cache_free_blocks):
+            self.state_indices[request_size:request_size +
+                               padding_size] = torch.tensor(
+                                   self.mamba_cache_free_blocks[:padding_size],
+                                   dtype=self.state_indices.dtype,
+                                   pin_memory=True).to(
+                                       self.state_indices.device,
+                                       non_blocking=True)
```
Comment (yuantailing) [inline tensorrt_llm/_torch/pyexecutor/mamba_cache_manager.py]: Can this branch be dropped? It would be more robust if we have less branches.

**Review by nv-guomingz — APPROVED** [review]

**Review by syuoni — APPROVED** [review]
Approved for the cuda graph and test changes.
